test/core/rules/L001_test.py

```diff
@@ -0,0 +1,25 @@
+# flake8: noqa
+"""Tests for L001."""
+from .std_test import rules__test_helper
+
+
+test_cases = {
+    "pass": [],
+    "fail": [
+        {
+            "query":
+"""
+SELECT 1     
+""",
+            "fixed":
+"""
+SELECT 1
+""",
+            "configs": None
+        }
+    ],
+}
+
+
+def test__rule_L001():
+    rules__test_helper("L001", test_cases)
```

Comment:
Right now, this would make one _pytest_ test case for all of the different cases in the structure.

I think a pattern more like:
```
@pytest.mark.parametrize(
    "test_case",
    [
        PassingRuleTestCase("SELECT 1"),
        FailingRuleTestCase("SELECT 1    \n", fixed="SELECT 1\n"),
        FailingRuleTestCase("...", fixed="...", configs={"whatever": "we need"}),
    ]
)
def test__rule_L001(test_case):
    run_rules_test_case("L001", test_case)
```

...might be more useful. In this way we define the rules directly as test parameters rather than setting up a structure which we then deconstruct. Each of the cases is also defined separately, and therefore we could see failures separately.

Thoughts?

Reply:
Definitely 100% agree we should have a pytest test per test case. My only qualm is that I'd like a way to make multiline test cases easier to both write and read. Lots of the rules now need pretty complex queries to fully test them, with huge parameterisations I think it gets messy with triple quote strings (as I do with the dict one in the draft). I wonder if a solution that creates a class per test case, where the class is an overridden named tuple or dataclass could work well...